[Relay][Op] Clip

joshpoll opened the pull request:
This PR registers the clip operator. I was wondering if I should put additional constraints on the attributes, though. `a_min` and `a_min` should be rank 0 tensors, which is not reflected in the current type relation. Additionally, should their data types match that of the input tensor `a`?

tqchen commented on src/relay/op/tensor/unary.cc:
```diff
@@ -78,5 +78,37 @@ RELAY_REGISTER_OP("concat")
+// Clip
+struct ClipAttrs : public tvm::AttrsNode<ClipAttrs> {
+  // TODO: should I enforce these be 0 dim?
+  Constant a_min;
```
just use double as they are constant and double is enough to represent most numbers we care about

tqchen commented on python/tvm/relay/op/tensor.py:
```diff
@@ -244,3 +244,35 @@ def concat(*args):
+    ----------
+    a : relay.Expr
+        The input tensor.
+    a_min : relay.Constant
```
let us just allow float

tqchen commented on python/tvm/relay/op/tensor.py:
```diff
@@ -244,3 +244,35 @@ def concat(*args):
+    ----------
+    a : relay.Expr
+        The input tensor.
+    a_min : relay.Constant
```
Although relay.Constant is more general in this case, let us just use float and make a_min and a_max attributes for now, this is the most common usecase of clip

joshpoll replied:
Do you mean that `a_min` and `a_max` should be python attributes? That deviates from numpy's function signature and also makes it impossible to enforce that `a_min` and `a_max` must be provided.

tqchen replied:
I mean make a_min and a_max as part of Attrs, see e.g. https://github.com/dmlc/tvm/blob/master/include/tvm/relay/attrs/transform.h#L17

joshpoll commented:
@tqchen I addressed your feedback. My only lingering concern is that the data type of the tensor is changed by clip. I could write a custom type relation to express this.

tqchen commented:
Why should data type of the tensor be changed by clip? I think we can enforce it to be the same as input for now,

joshpoll commented:
Sorry. I was mistaken. I looked more closely at the clip implementation and I see that the bounds are cast to the input tensor's dtype.
